Postgres: Information schema add user defined type

cliedeman opened the pull request.

a8m commented on dialect/sql/schema/postgres.go:
```diff
@@ -87,7 +87,7 @@ func (d *Postgres) setRange(ctx context.Context, tx dialect.Tx, t *Table, value
 	rows := &sql.Rows{}
 	query, args := sql.Dialect(dialect.Postgres).
-		Select("column_name", "data_type", "is_nullable", "column_default").
+		Select("column_name", "data_type", "is_nullable", "column_default", "udt_name").
```
Do you need this column for future PRs? because you don't use it atm.

cliedeman replied:
Yes. it used in any future feature that needs access to the custom type. E.g. composite types or enums

a8m commented on dialect/sql/schema/postgres.go:
```diff
@@ -214,16 +214,19 @@ const maxCharSize = 10 << 20
-		defaults sql.NullString
+		nullable        sql.NullString
+		defaults        sql.NullString
+		userDefinedType sql.NullString
```
Give it a shorter name please (e.g. `userType` or `udt`).

a8m commented on dialect/sql/schema/postgres.go:
```diff
@@ -214,16 +214,19 @@ const maxCharSize = 10 << 20
 		c.Nullable = nullable.String == "YES"
 	}
 	switch c.typ {
+	case "USER-DEFINED":
+		c.Type = field.TypeOther
```
Don't you want to move it to line 255? Or you plan to add more logic here.

cliedeman replied:
nice catch, no extra logic for now

a8m reviewed (commented):
Thanks for adding this @cliedeman

a8m commented on dialect/sql/schema/postgres_test.go:
```diff
@@ -192,10 +192,10 @@ func TestPostgres_Create(t *testing.T) {
-					WillReturnRows(sqlmock.NewRows([]string{"column_name", "data_type", "is_nullable", "column_default"}).
-						AddRow("id", "bigint", "NO", "nextval('users_colname_seq'::regclass)"))
+					WillReturnRows(sqlmock.NewRows([]string{"column_name", "data_type", "is_nullable", "column_default", "udt_name"}).
+						AddRow("id", "bigint", "NO", "nextval('users_colname_seq'::regclass)", "NULL"))
```
```suggestion
						AddRow("id", "bigint", "NO", "nextval('users_colname_seq'::regclass)", "int4"))
```

a8m commented on dialect/sql/schema/postgres_test.go:
```diff
@@ -229,20 +229,20 @@ func TestPostgres_Create(t *testing.T) {
-						AddRow("macaddr", "macaddr", "YES", "NULL").
-						AddRow("macaddr8", "macaddr8", "YES", "NULL"))
+					WillReturnRows(sqlmock.NewRows([]string{"column_name", "data_type", "is_nullable", "column_default", "udt_name"}).
+						AddRow("id", "bigint", "NO", "NULL", "NULL").
+						AddRow("name", "character varying", "YES", "NULL", "NULL").
+						AddRow("uuid", "uuid", "YES", "NULL", "NULL").
+						AddRow("created_at", "date", "NO", "CURRENT_DATE", "NULL").
+						AddRow("updated_at", "timestamp", "YES", "NULL", "NULL").
+						AddRow("deleted_at", "date", "YES", "NULL", "NULL").
+						AddRow("text", "text", "YES", "NULL", "NULL").
+						AddRow("cidr", "cidr", "NO", "NULL", "NULL").
+						AddRow("inet", "inet", "YES", "NULL", "NULL").
+						AddRow("macaddr", "macaddr", "YES", "NULL", "NULL").
+						AddRow("macaddr8", "macaddr8", "YES", "NULL", "NULL"))
```
```suggestion
						AddRow("id", "bigint", "NO", "NULL", "int8").
						AddRow("name", "character varying", "YES", "NULL", "varchar").
						AddRow("uuid", "uuid", "YES", "NULL", "uuid").
						AddRow("created_at", "date", "NO", "CURRENT_DATE", "date").
						AddRow("updated_at", "timestamp", "YES", "NULL", "timestamptz").
						AddRow("deleted_at", "date", "YES", "NULL", "date").
						AddRow("text", "text", "YES", "NULL", "text").
						AddRow("cidr", "cidr", "NO", "NULL", "cidr").
						AddRow("inet", "inet", "YES", "NULL", "inet").
						AddRow("macaddr", "macaddr", "YES", "NULL", "macaddr").
						AddRow("macaddr8", "macaddr8", "YES", "NULL", "macaddr8"))
```

a8m commented on dialect/sql/schema/postgres_test.go:
```diff
@@ -270,12 +270,12 @@ func TestPostgres_Create(t *testing.T) {
-						AddRow("id", "bigint", "NO", "NULL").
-						AddRow("name", "character", "YES", "NULL").
-						AddRow("doc", "jsonb", "YES", "NULL"))
+					WillReturnRows(sqlmock.NewRows([]string{"column_name", "data_type", "is_nullable", "column_default", "udt_name"}).
+						AddRow("id", "bigint", "NO", "NULL", "NULL").
+						AddRow("name", "character", "YES", "NULL", "NULL").
+						AddRow("doc", "jsonb", "YES", "NULL", "NULL"))
```
```suggestion
					WillReturnRows(sqlmock.NewRows([]string{"column_name", "data_type", "is_nullable", "column_default", "udt_name"}).
						AddRow("id", "bigint", "NO", "NULL", "int8").
						AddRow("name", "character", "YES", "NULL", "bpchar").
						AddRow("doc", "jsonb", "YES", "NULL", "jsonb"))
```

a8m commented on dialect/sql/schema/postgres_test.go:
```diff
@@ -304,12 +304,12 @@ func TestPostgres_Create(t *testing.T) {
-						AddRow("id", "bigint", "NO", "NULL").
-						AddRow("name", "character", "YES", "NULL").
-						AddRow("doc", "jsonb", "YES", "NULL"))
+					WillReturnRows(sqlmock.NewRows([]string{"column_name", "data_type", "is_nullable", "column_default", "udt_name"}).
+						AddRow("id", "bigint", "NO", "NULL", "NULL").
+						AddRow("name", "character", "YES", "NULL", "NULL").
+						AddRow("doc", "jsonb", "YES", "NULL", "NULL"))
```
```suggestion
					WillReturnRows(sqlmock.NewRows([]string{"column_name", "data_type", "is_nullable", "column_default", "udt_name"}).
						AddRow("id", "bigint", "NO", "NULL", "int8").
						AddRow("name", "character", "YES", "NULL", "bpchar").
						AddRow("doc", "jsonb", "YES", "NULL", "jsonb"))
```

a8m commented on dialect/sql/schema/postgres_test.go:
```diff
@@ -336,11 +336,11 @@ func TestPostgres_Create(t *testing.T) {
-						AddRow("id", "bigint", "NO", "NULL").
-						AddRow("name", "character", "YES", "NULL"))
+					WillReturnRows(sqlmock.NewRows([]string{"column_name", "data_type", "is_nullable", "column_default", "udt_name"}).
+						AddRow("id", "bigint", "NO", "NULL", "NULL").
+						AddRow("name", "character", "YES", "NULL", "NULL"))
```
```suggestion
						AddRow("id", "bigint", "NO", "NULL", "int8").
						AddRow("name", "character", "YES", "NULL", "bpchar"))
```

a8m commented on dialect/sql/schema/postgres_test.go:
```diff
@@ -367,11 +367,11 @@ func TestPostgres_Create(t *testing.T) {
-						AddRow("id", "bigint", "NO", "NULL").
-						AddRow("name", "character", "YES", "NULL"))
+					WillReturnRows(sqlmock.NewRows([]string{"column_name", "data_type", "is_nullable", "column_default", "udt_name"}).
+						AddRow("id", "bigint", "NO", "NULL", "NULL").
+						AddRow("name", "character", "YES", "NULL", "NULL"))
```
```suggestion
						AddRow("id", "bigint", "NO", "NULL", "int8").
						AddRow("name", "character", "YES", "NULL", "bpchar"))
```

a8m commented on dialect/sql/schema/postgres_test.go:
```diff
@@ -398,11 +398,11 @@ func TestPostgres_Create(t *testing.T) {
-						AddRow("id", "bigint", "NO", "NULL").
-						AddRow("name", "character", "YES", "NULL"))
+					WillReturnRows(sqlmock.NewRows([]string{"column_name", "data_type", "is_nullable", "column_default", "udt_name"}).
+						AddRow("id", "bigint", "NO", "NULL", "NULL").
+						AddRow("name", "character", "YES", "NULL", "NULL"))
```
```suggestion
						AddRow("id", "bigint", "NO", "NULL", "int8").
						AddRow("name", "character", "YES", "NULL", "bpchar"))
```

a8m commented on dialect/sql/schema/postgres_test.go:
```diff
@@ -428,11 +428,11 @@ func TestPostgres_Create(t *testing.T) {
-						AddRow("id", "bigint", "NO", "NULL").
-						AddRow("name", "character", "YES", "NULL"))
+					WillReturnRows(sqlmock.NewRows([]string{"column_name", "data_type", "is_nullable", "column_default", "udt_name"}).
+						AddRow("id", "bigint", "NO", "NULL", "NULL").
+						AddRow("name", "character", "YES", "NULL", "NULL"))
```
```suggestion
						AddRow("id", "bigint", "NO", "NULL", "int8").
						AddRow("name", "character", "YES", "NULL", "bpchar"))
```

a8m commented on dialect/sql/schema/postgres_test.go:
```diff
@@ -458,11 +458,11 @@ func TestPostgres_Create(t *testing.T) {
-						AddRow("id", "bigint", "NO", "NULL").
-						AddRow("name", "character", "NO", "NULL"))
+					WillReturnRows(sqlmock.NewRows([]string{"column_name", "data_type", "is_nullable", "column_default", "udt_name"}).
+						AddRow("id", "bigint", "NO", "NULL", "NULL").
+						AddRow("name", "character", "NO", "NULL", "NULL"))
```
```suggestion
						AddRow("id", "bigint", "NO", "NULL", "int8").
						AddRow("name", "character", "NO", "NULL", "bpchar"))
```

a8m commented on dialect/sql/schema/postgres_test.go:
```diff
@@ -488,11 +488,11 @@ func TestPostgres_Create(t *testing.T) {
-						AddRow("id", "bigint", "NO", "NULL").
-						AddRow("age", "bigint", "NO", "NULL"))
+					WillReturnRows(sqlmock.NewRows([]string{"column_name", "data_type", "is_nullable", "column_default", "udt_name"}).
+						AddRow("id", "bigint", "NO", "NULL", "NULL").
+						AddRow("age", "bigint", "NO", "NULL", "NULL"))
```
```suggestion
						AddRow("id", "bigint", "NO", "NULL", "int8").
						AddRow("age", "bigint", "NO", "NULL", "int8"))
```

a8m commented on dialect/sql/schema/postgres_test.go:
```diff
@@ -518,11 +518,11 @@ func TestPostgres_Create(t *testing.T) {
-						AddRow("id", "bigint", "NO", "NULL").
-						AddRow("age", "bigint", "NO", "NULL"))
+					WillReturnRows(sqlmock.NewRows([]string{"column_name", "data_type", "is_nullable", "column_default", "udt_name"}).
+						AddRow("id", "bigint", "NO", "NULL", "NULL").
+						AddRow("age", "bigint", "NO", "NULL", "NULL"))
```
```suggestion
						AddRow("id", "bigint", "NO", "NULL", "int8").
						AddRow("age", "bigint", "NO", "NULL", "int8"))
```

a8m commented on dialect/sql/schema/postgres_test.go:
```diff
@@ -548,11 +548,11 @@ func TestPostgres_Create(t *testing.T) {
-						AddRow("id", "bigint", "NO", "NULL").
-						AddRow("age", "bigint", "NO", "NULL"))
+					WillReturnRows(sqlmock.NewRows([]string{"column_name", "data_type", "is_nullable", "column_default", "udt_name"}).
+						AddRow("id", "bigint", "NO", "NULL", "NULL").
+						AddRow("age", "bigint", "NO", "NULL", "NULL"))
```
```suggestion
						AddRow("id", "bigint", "NO", "NULL", "int8").
						AddRow("age", "bigint", "NO", "NULL", "int8"))
```

a8m commented on dialect/sql/schema/postgres_test.go:
```diff
@@ -602,12 +602,12 @@ func TestPostgres_Create(t *testing.T) {
-						AddRow("age", "bigint", "NO", "NULL").
-						AddRow("score", "bigint", "NO", "NULL"))
+					WillReturnRows(sqlmock.NewRows([]string{"column_name", "data_type", "is_nullable", "column_default", "udt_name"}).
+						AddRow("id", "bigint", "NO", "NULL", "NULL").
+						AddRow("age", "bigint", "NO", "NULL", "NULL").
+						AddRow("score", "bigint", "NO", "NULL", "NULL"))
```
```suggestion
						AddRow("id", "bigint", "NO", "NULL", "int8").
						AddRow("age", "bigint", "NO", "NULL", "int8").
						AddRow("score", "bigint", "NO", "NULL", "int8"))
```

a8m commented on dialect/sql/schema/postgres_test.go:
```diff
@@ -623,11 +623,11 @@ func TestPostgres_Create(t *testing.T) {
-						AddRow("id", "bigint", "NO", "NULL").
-						AddRow("score", "bigint", "NO", "NULL"))
+					WillReturnRows(sqlmock.NewRows([]string{"column_name", "data_type", "is_nullable", "column_default", "udt_name"}).
+						AddRow("id", "bigint", "NO", "NULL", "NULL").
+						AddRow("score", "bigint", "NO", "NULL", "NULL"))
```
```suggestion
						AddRow("id", "bigint", "NO", "NULL", "int8").
						AddRow("score", "bigint", "NO", "NULL", "int8"))
```

a8m commented on dialect/sql/schema/postgres_test.go:
```diff
@@ -664,11 +664,11 @@ func TestPostgres_Create(t *testing.T) {
-						AddRow("id", "bigint", "YES", "NULL").
-						AddRow("name", "character", "YES", "NULL"))
+					WillReturnRows(sqlmock.NewRows([]string{"column_name", "data_type", "is_nullable", "column_default", "udt_name"}).
+						AddRow("id", "bigint", "YES", "NULL", "NULL").
+						AddRow("name", "character", "YES", "NULL", "NULL"))
```
```suggestion
						AddRow("id", "bigint", "YES", "NULL", "int8").
						AddRow("name", "character", "YES", "NULL", "bpchar"))
```

a8m commented on dialect/sql/schema/postgres_test.go:
```diff
@@ -730,10 +730,10 @@ func TestPostgres_Create(t *testing.T) {
-					WillReturnRows(sqlmock.NewRows([]string{"column_name", "data_type", "is_nullable", "column_default"}).
-						AddRow("id", "bigint", "YES", "NULL"))
+					WillReturnRows(sqlmock.NewRows([]string{"column_name", "data_type", "is_nullable", "column_default", "udt_name"}).
+						AddRow("id", "bigint", "YES", "NULL", "NULL"))
```
```suggestion
						AddRow("id", "bigint", "YES", "NULL", "int8"))
```

a8m reviewed (approved):
Thanks for adding this @cliedeman.

I added multiple change suggestions for test cases, I prefer to use the real values in mocks.

cliedeman commented:
@a8m I didn't merge in the unit test changes because those nulls are expected. The only time udt will not be null is when using a type created with `CREATE TYPE`. I have added another unit test to cover this case and added an integration test.

The pr will now allow migrations with custom schema types whem the type does not change

a8m commented:
> @a8m I didn't merge in the unit test changes because those nulls are expected. The only time udt will not be null is when using a type created with `CREATE TYPE`. I have added another unit test to cover this case and added an integration test.
> 
> The pr will now allow migrations with custom schema types when the type does not change

I just checked it on PostgreSQL 10/11/12 and I got this:

![image](https://user-images.githubusercontent.com/7413593/103169991-4e653480-4849-11eb-88a8-ff03a1c5d773.png)

cliedeman commented:
@a8m nothing else to say except that I am _completely_ wrong. Will update the unit tests

a8m commented:
Thanks for adding this @cliedeman 😃
